### ISSUE TITLE
Future CQ Development

### ISSUE THREAD

**Issue by jmwright** [dcowden/cadquery#40]
There has been some informal discussion about making CQ work with multiple kernels, and I'd like to try to get a conversation started on the nuts and bolts of doing that.

**Comment by jmwright** [comment]
I'm interested in experimenting with making units easy in CQ. This may just take the form of some best practice documentation, but I'm not sure yet. This library came across Twitter today and looks promising. https://github.com/dcowden/cadquery

**Comment by emdash** [comment]
Did you mean to paste a different link? That's just the cadquery page...

**Comment by jmwright** [comment]
Yeah, thanks for catching that. http://pint.readthedocs.org/en/0.6/

**Comment by jmwright** [comment]
This isn't really a future development item, but I wasn't sure where else to put it. A thought occurred to me the other day that the following code should work fine, and not surprisingly, it did.
```python
import cadquery
from Helpers import show

length = 2.0
height = 1.0
thickness = 1.0

result = eval('cadquery.Workplane("XY").box(length, height, thickness)')

show(result)
```
If used carefully and not abused, this could lead to some interesting techniques like dynamically generating parts of a model's code at runtime. I would not promote the use of eval for everyday scripts, but it could open some doors on challenging modelling/analysis situations later on.

**Comment by emdash** [comment]
I think this is a good example of the sort of turing-complete trouble you will get into because you've chosen an imperative language for scripting rather than a purely-functional one, or a domain-specific declarative one.

You literally cannot predict all valid usage.

The advantage of python is that you have a lot of library support. The disadvantage is that it's impossible to analyze a script in the general case. It's the same problem that other analysis tools like coverity have: you can spot a particular instance of a construct, but you can't prove that a construct isn't used, that a certain class of error doesn't exist, or even that a script terminates.

I completely understand why the OpenScad folks designed their own language. I actually think they had the right idea, but the OpenScad implementation has some frustrating behavior. Some are due to it's being a CSG kernal, rather than BREP. But others are due to silly, avoidable problems, like the semantics of variable scoping within loops (which is entirely nonsensical, and I consider it to be a bug not a feature. They could have implemented proper lexical scoping and it's a mystery to me why they did .

At this point I am not sure whether the Turing-complete / imperative nature of python is an asset or a liability. It lets me get away with some things, but I feel like most of them could be covered with a targeted set of declarative features. And I might well be willing to trade Turing-completeness in exchange for various guarantees: i.e. that a script can be evaluated in a reasonable amount of time, that it can be analyzed for useful properties (like parameter names).

In other words, I'm not entirely on board with the fundamental CadQuery assumption: that pure python is the right language to express CAD models. There are various alternatives I would consider. On the one hand, a completely domain-specific language (like OpenScad's) that has syntactic support for CAD operations. On the other hand, a restricted subset of python which allows for meaningful static analysis. Even some kind of XML dialect wouldn't be out of the question in my book.

Anyways, apologies for the essay.

What CadQuery has going for it right now is that it's a concise programmatic interface to a BREP kernel, with some really useful innovations, like the concept of selectors. There are many ways you could go about doing that. Python has some things going for it, but there are other languages out there.

**Comment by dcowden** [comment]
The advantages of using a real language far outweight the few problems
associated with problems like the above.

Sure you end up with a language where you can concoct a bunch of things
that you can try to but don't work-- but who cares-- that's not practically
worse than simply not supporting them, given that the alternative is to
design a complete other language, or, to use a language that has no native
bindings available for the open open source BREP kernel out there.

At this point I am not sure whether the Turing-complete / imperative nature
of python is an asset or a liability

I am. Python is clearly the right choice here. If you want to use OCC, you
get python, C, or C++, or a custom designed language. There's not enough
value in building ( and maintaining ) your own language just to avoid a few
weird scenarios that _someone_  _might_ try, and dont work anyway.

**Comment by emdash** [comment]
I think that if you ever want to support interactive updates to the code, where interactive changes cause live code updates, this won't be possible in general if the input language is turing complete. It might be possible if the input language is declarative.

Let's try a simple example:

show(Workplane("XY").circle(3).extrude(10))

Let's say I want to select the top face of the cylinder with the mouse and drag it upward 3mm, with the result that the code now says:

show(Workplane("XY").circle(3).extrude(13))

In something simple like this, you could work backward from the top face to find out it's created by a call to extrude, and then use some internal book-keeping to point you at the source code line and know that we are adding 3mm to the first parameter to the call to extrude. But if we instead had:

 show(Workplane("XY").circle(3).extrude(HEIGHT))

It gets less clear. Probably the expression be transformed to:

 show(Workplane("XY").circle(3).extrude(HEIGHT + 3))

Which is still doable, but what if instead we have:

while x > 0:

**Comment by dcowden** [comment]
If the use case is updating a script from a 3d view of a model, the
complexities of updating the variable in the script are the least of your
worries. You wouldn't have any straightforward way to know that a
particular mouse movement corresponds to any particular parameter. In fact,
in the general case that will not be the case. Regardless of what language
you use, the best you could do is to identify a feature with a click, and
then let the user change the inputs for that feature. there's no way to
know which inputs to change just based on a click-- unless the click is
interacting with a text field that controls the variable's value. then, its
easy-- you just re-run the script with the new variable value.

I'd be happy if we had the ability to select, by clicking, a given feature.
that's already sufficiently difficult, but doesn't begin to associate which
input(s) generated that feature.

**Comment by emdash** [comment]
Source highlighting would definitely be the first step.

It's really a book-keeping problem. On the one hand you have your feature tree, which has to do book-keeping already in order to implement all the selectors. Each node in the feature tree needs to know the which faces it produced. Then you can construct your reverse mapping from vertex / edge / face to feature nodes. So when the user selects one, you know which nodes in the feature tree are of interest.

To implement your selection highlighting, you also need another reverse mapping from feature nodes to source lines. There needs to be some way to capture the source code line in the script which produces a feature. In pure python, that will be the tricky / ugly, but I believe it is possible. You would probably have to introspect the call stack, but this could maybe be factored into a function decorator, such that any method that needs to know what source file / line it came from gets that passed in as a keyword argument. Either that or PDB or some other profiling tool could be employed that would analyze the call graph after the script runs.

But basically when you call a function like cq.Box(), or cq.Union(), the idea is that it would implicitly capture it's call site location and store it.

There might be multiple results for a given vertex / edge / face. If that happens, I guess you just highlight all the relevant lines. You might even consider highlighting all the way up the call stack, in the event that the call occurs in something other than a module-level statement.

**Comment by emdash** [comment]
Actually, you probably also want to be able to go the other way: you should be able to see which features were produced by a given line of code. But if you can do the one, you can also do the other.

Another useful feature, which I may have mentioned already, but the viewer should have an "interactive selector builder". I.e. I can type a selector expression into a bar, and it interactively highlights the features that match. And puts it in the clipboard.

What would be really cool is if you could go the other way: I click on some feature, and it produces a list of selector chains which will match that feature. My gut tells me this it is computationally intractable to find an optimal selector chain. But maybe if it could be useful as a short-hand for the "in the direction-of" or "nearest-to-point" selectors.

**Comment by emdash** [comment]
For code updates, you would need not just line numbers, but the AST for the script. Instead of features mapping to line numbers, the feature tree would map to the AST node which produced the feature (this is much easier to handle with a declarative language. There will be cases in python where a useful syntax node can't be identified).

What interactive model changes would do is update the feature tree, which would in turn update the AST, and then the code would be "rendered" into the editor from the AST. I have implemented this kind of thing before, though not for a general-purpose programming language. Preserving the arbitrary formatting of the input would be a challenge in python (I don't think the python parser module preserves white space, unlike a conforming XML parser).

As for deciding what to change, when there's multiple possibilities, you can allow the user to choose among the available ones. Consider:

 show(Workplane("XY").rect(width, width * 2).extrude(5).edges("|Z").fillet(6).translate((0, 0, 1))

So, If I understand correctly, the feature tree / graph looks like this (or the reversed version):

- translate
  - fillet
    - extrude
      - rect
        - workplane

Now imagine, you select a wire that was created by the fillet operation. Since a fillet node is in the selection, the following options should appear:
- the move command (provided by the translate node)
- fillet radius (provided by the fillet node)
- extrusion depth (provided by the extrude node)
- (read only) rect width
- (read only) rect height

I.e. what we can manipulate from the viewer are any arguments to the feature tree that map to literals in the source code.

**Comment by dcowden** [comment]
that feature tree is definitely what i had in mind, except for two things:

(1) i think we may not want to create a new feature for things like
rotate/translate that we know do not produce any different 3d features.
(ie, no new feature for translate ),
(2) not sure we want to have features for 2d constructs ( rect)-- but not
sure about that yet.

My plan was to mirror the way Solidworks does it, which seems to work
pretty well.  In SW, you have an object in the feature tree for a workplane
( because you need to specify a workplane in order to extrude), and you
have objects for newly created 3d solids. but you dont have objects for
each line/arc/curve/rect on the plane, and you dont have a new feature when
you move things around.



I'd be happy to achieve selecting features by clicking as a start. IE, you
click on one vertex/edge/wire/face and select all of the other
wire/edge/faces that were produced in the same feature.

**Comment by emdash** [comment]
Conceptually, I think a rotate or a translate produces a new solid, no? In which case it should appear in the feature tree.

I kindof don't like the notion that basic transformations mutate the node in the feature tree rather than produce a new object.

If I say the following:
 plate = Workplane("XY").box(l, w, h)
 hole = Workplane("XY").polygon(6, size).extrude(depth)
 plate = plate.cut(hole.translate(leftHoleLocation))
 plate = plate.cut(hole.translate(rightHoleLocation))

What I want is clearly a plate with two holes in it. Currently, you only get one hole (the right one). To work around this, I have to make a "hole function" so I can create a multiple instances of the hole to cut from. I shouldn't have to do that, though. It's especially problematic with loops.

As for whether 2D operations should produce features, I feel like they should exist in some form. At the very least, it would be useful to see unconsumed points and wires hilighted in the output. This could indicate a bug. But it could also be a useful visualization tool.

**Comment by jmwright** [comment]
Here's a link to the [ast module](https://docs.python.org/2/library/ast.html) in Python, which I'm thinking would help with source highlighting. Line numbers could be a fall-back when a useful syntax node couldn't be found for a feature.

**Comment by emdash** [comment]
Yeah, I had originally envisioned using the ast and compile modules. Basically, if your input language looks enough like python syntactically, you don't have to write your own parser. It becomes trivial to do validation and to identify line numbers for a given bit of syntax.

I've written some code that makes heavy use of the Ast module and have a good idea how to work with it. If you want to restrict that certain things not be done (like forbidding the creation of parameters within functions and loops), an Ast pass makes it pretty easy to do.

Crazier things are possible. Once you have the AST, you are free to change the semantics as you see fit. You could conceivably define functional / declarative semantics that would allow you to accurately map features to code, and to do interactive code updates.

**Comment by jmwright** [comment]
It seems like for maximum benefit the feature tree, this AST work, and the code/feature highlighting in the GUI all need to be done in concert. They don't have to be tightly coupled, just developed in parallel so that they can inform each other's development process.

**Comment by emdash** [comment]
I think the feature tree comes first. At least, that's the part that in my mind I have the least understanding of. And part of me thinks that there may need to be two or three parallel trees:

1) The input tree, which defines the model in terms of BREP primatives and operations. This is what results directly from evaluating the python code.
2) The output tree is what we actually send to PyOCC.
3) Possibly an intermediate tree between 1) and 2) which would track the intermediate objects that result from processing the input tree. May be needed for implementing some of the selectors, for example.

Also, the input tree is really a graph. A given node can appear in multiple places.

**Comment by jmwright** [comment]
I've started experimenting with possible components for the new CQ GUI, and want to get some feedback from the ground up. I've created a [short video](http://youtu.be/Z54LUHjwEgU) where I do a test drive of the 3D viewer that's embedded with PythonOCC so that all of you can comment on how it works and looks. As @emdash has pointed out in this discussion, solid 3D viewer interaction is a must-have in CAD software.

The video is very poor quality. I did it in one take with no prep. The purpose of the video was to let the CQ community see the viewer and comment on it, not make quality content.

The first comment I have is that the right click and drag to zoom seems redundant (already taken care of with the mouse wheel). I would also like to reserve the right click for the code interaction features that we hope will be an integral part of the new CQ GUI. I'll wait to make more comments to give others a chance to chime in.

If you want to test drive this PythonOCC 3D viewer yourself, you can use the [pythonocc_oce_setup](https://github.com/jmwright/pythonocc_oce_setup) script on any Ubuntu 14.04 flavor. Then you can run the example from the directory that I show in the video.

**Comment by dcowden** [comment]
I watched the video-- that viewer would certainly be good enough to start
with. I agree that right click would be good to use for menus and stuff.

I'm interested in the edge/face selection. If it does that, it could be
really useful.  If we tracked a map of what CQ features produced which
edges/faces, we could use edge/face selection to select subfeatures in the
feature tree, and maybe even highlight the code that produced that feature.

**Comment by jmwright** [comment]
Yeah, the face/edge selection caught my attention too. The source code also seems to support vertex selection, but it doesn't work in the viewer. Shouldn't be hard to add that support though.

**Comment by emdash** [comment]
This is really exciting. I am going to take a look when I get home. :)

**Comment by jmwright** [comment]
Started an [issue on pythonocc-core](https://github.com/tpaviot/pythonocc-core/issues/89) about bundling PythonOCC and OCE within an application. That way I can share any work being done on the new CQ GUI mock-up more easily.

**Comment by jmwright** [comment]
I don't have OCE embedded, but I put together a proof-of-concept UI using PyQode (the FreeCAD module's editor), the OCCViewer, and PySide.

![cq_experimental_gui](https://cloud.githubusercontent.com/assets/1015439/6716546/9ec3aa0e-cd7c-11e4-8386-acc65529f662.png)

I'm probably only 4 to 8 hours away from having a decent GUI that will execute straight PythonOCC code and render it in the viewer. That state of the GUI would allow us to experiment with the PythonOCC port of CQ easily, but probably won't do us much good until we have time to work on that port.

**Comment by emdash** [comment]
This is actually really good news. When the GUI is functional, I can start
tinkering with PyOCC in my off-time. The goal would just be to get a feel
for the library at first, and then eventually contribute to the task of
re-implementing the CQ API.

While I like the CQ concept, the shortcomings of the current implementation
(the lack of polygon offsetting for proper clearances, in particular)
really limit what I can use CQ for. At least this way, I can get back to
constructing models programmatically, even if the API is a little verbose.
I think I would learn a lot from doing at least a few projects directly in
PyOCC.

One thing that I really need is a way to render translucent objects. I've
found that this is really important for evaluating model features,
especially when you are trying to create interlocking parts, and
thin-shelled solids. It can also be a useful tool for debugging
programmatic CAD. I.e. you can have translucent representations of
intermediate shapes, like OpenSCAD.

--Brandon

**Comment by jmwright** [comment]
> One thing that I really need is a way to render translucent objects.

@emdash The current release of the CQ module for FreeCAD allows you to set and render part colors and transparency. The OCCViewer is supposed to support this as well

**Comment by emdash** [comment]
Right, what I'm saying is that making sure this works in the OCC viewer
would be key.

**Comment by jmwright** [comment]
OK, got it. Thanks.

**Comment by emdash** [comment]
It's been a while since there was an update on this issue.. How are things going?

**Comment by dcowden** [comment]
Hi there:

In short, not much.  I've been pulled off onto other projects for a bit.
 @jmwright has been making progress on a pythonOCC branch of CQ, but its
also been slow going.

**Comment by emdash** [comment]
Is his work public? If so, I wouldn't mind a link to it.

**Comment by dcowden** [comment]
sure, hes working on the pythonocc branch of cadquery:

https://github.com/dcowden/cadquery/tree/pythonocc

**Comment by emdash** [comment]
Thanks, I'll take a look when I get home :)

--Brandon

**Comment by jmwright** [comment]
That branch is me experimenting with a PythonOCC implementation that lives along side the FreeCAD implementation. It's not an attempt to implement any of our more advanced ideas, it's just a direct port. I got enough experience with PythonOCC to contribute a port of the classic OCC bottle example to pythonocc-core, but just got bogged down after that. With @dcowden busy, I'm the main contributor right now and don't have the free time to keep the current version moving forward, and do the PythonOCC impl too. Here's some more if our conversation in this -  https://github.com/dcowden/cadquery/issues/89

I was building the TestPythonOCC test up as I wrote the implementation, so that would be the place to start. The test covers pythonocc_impl/geom only for right now.

**Comment by emdash** [comment]
OK. I am still interested in potentially contributing, but I am wondering
what the next stop on the roadmap would be. OCC has a ridiculously verbose
API, I've noticed. I have this feeling that really only a small fraction of
it is relevant to a project like CadQuery, and I'd be curious to read any
notes you might have on it.

--Brandon

**Comment by dcowden** [comment]
I support the pythonOCC direction. Yes the API is very verbose. But its
attractive because its the only<python> game in town that provides
low-level access to all of the features of OpenCascade-- which is the most
capable O/S kernel available.

FreeCAD (FC) is limiting. Though existing CQ is decent, many of the things
we would need to do next are not available in FC. @jmwright has spent some
time asking them for changes, and they are not receptive at all to
contributions and new ideas. They border on downright uncooperative.  In
short, we've both concluded that if CQ needs to grow, it will be very slow
as long as FC is underneath.  PythonOCC, in contrast, is supported by two
folks who are super helpful and responsive, and willing to accept
contributions. Thomas and Jelle are both super excited, and would accept
contributes to pythonOCC from us in a second. I've even considered making
CQ part of pythonOCC.

Most of the other possible directions involve using one of many CSG
kernels. My gut feeling is that these will ultimately make matters even
worse than using pythonOCC, because things that need to be easy, like
filleting and extrusion, become a lot more work in CSG.  There's a reason
why all of the top professionally used CAD packages are BREP based and not
CSG based.

Its not ideal, but i think its the best strategy.

**Comment by emdash** [comment]
Just wrapping up a project that should give me a little more time to tinker with CadQuery. I got the pythonocc_oce_setup script to build on an ubuntu VM. But I'm a little confused about what needs to happen next. It sounds like @jmwright has a different branch that I should be messing with?

**Comment by jmwright** [comment]
@emdash - Check out the pythonocc branch. https://github.com/dcowden/cadquery/tree/pythonocc?files=1

It's got a cadquery/pythonocc_impl directory, and I started in geom.py. It's pretty rigged in its current state. I was just building and running the tests manually as I went. There's no integration with the wider library yet because it will throw all sorts of import errors because the implementation isn't complete.

Let me know if you have questions.

**Comment by emdash** [comment]
Just one. Are you available on-line at all this week to help me get
bootstrapped on it?

--Brandon

**Comment by jmwright** [comment]
This week's pretty booked. I might be able to on Sunday evening.

**Comment by jmwright** [comment]
@emdash Did you ever get a chance to start experimenting with PythonOCC?

**Comment by emdash** [comment]
I kindof played with the viewer, but I haven't had time to get into PythonOCC itself.

**Comment by emdash** [comment]
I'll add that I finished a big project that I was doing, and now I have some more free time, and some lighter-weight projects, so this might be a relatively good time to try fiddling with it.

**Comment by jmwright** [comment]
Cool.

**Comment by jmwright** [comment]
This discussion has been moved to the new Google Group since it's long-running and not directly related to an issue in the codebase. https://groups.google.com/forum/?hl=en&nomobile=true#!topic/cadquery/AahXUjscM10

If you're a participant in this discussion, please join the group so that we can continue the discussion there.

**Comment by jmwright** [comment]
@emdash and @dcowden Can you guys give some more specifics on the keypad idea above? I'm experimenting with another GUI framework, and want to try that idea out. I use a 3D mouse when I do CAD, and I have the two buttons on it set up to switch to the isometric view and zoom to fit. I would like to have both of those operations still available with the keypad.

I was thinking that button 5 might be front view, then the other keys would be as follows. If you hit them once they just snap you to that view, but if you hold them down they could keep rotating the view. Maybe holding down on 5 or pushing it multiple times could toggle you between the front and back views?

8 - top view
4 - left side view
6 - right side view
2 - bottom view

7 - top left ISO
8 - top right ISO
1 - bottom left ISO
3 - bottom right ISO

There are also the mathematical operation keys. Maybe I can attach the zoom to fit operation to one of those?
